### DISCUSSION TITLE
Read environment variable in server only code in cloudflare-pages

### DISCUSSION THREAD

**Discussion by bondz** [remix-run/remix#3581] [Q&A]
#1285 addresses reading environment variables from context when using cloudflare-pages.

However, when you have other secrets to set, like client-secret, app-secret, etc for example when using social login, I am unable to  read environment variables.

I've tried reading from `globalThis` and I get the error

```
ReferenceError: A is not defined.
Attempted to access binding using global in modules.
You must use the 2nd `env` parameter passed to exported handlers/Durable Object constructors, or `context.env` with Pages Functions.
```

After setting `A=1` in `.env` and adding the `--binding` flag. However, I am able to read `context.A` in any loader in Remix.

**Comment by jgb-solutions** [comment]
Yeah it would be cool to read API_URL from a util or constant file. We will need to fetch those values outside of `loaders` and `actions` functions.

**Reply by ankumar3** [reply]
Did you guys find a solution for this? @jgb-solutions @bondz  or maybe a workaround to safely access the Env vars outside loader/action

**Reply by jgb-solutions** [reply]
@ankumar3 unfortunately no. Maybe there's a new way to do this but I kinda stopped working on the Cloudflare version and migrated to a full node app hosted on Vercel or Fly.io. Less hassle. Still fast.

**Reply by bondz** [reply]
Same. I ended up hosting my app on Netlify instead.

However, from this [article](https://blog.cloudflare.com/technology-behind-radar2/), looks like it should work but I don’t know how they achieved that.

**Comment by praagyajoshi** [comment]
Hello! 👋 

I'm in the same situation where I need to access `ENV` variables outside of a `loader` and `action`. Has there been any updates about it? Thank you!

**Comment by raviqqe** [comment]
If anyone still encounters this issue, we can use [the `import.meta.env` variable](https://vitejs.dev/guide/env-and-mode) at the build time now that Remix supports Vite. Just so you know, this is not a complete alternative because we cannot fetch environment variables dynamically. But I believe it should replace `process.env` in most of its use cases.

See also #7875.
